apm-agent-plugins/apm-logging-plugin/apm-log4j2-plugin/src/main/java/co/elastic/apm/agent/log4j2/correlation/CorrelationIdMapAdapter.java

```diff
@@ -0,0 +1,179 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.log4j2.correlation;
+
+import co.elastic.apm.agent.impl.GlobalTracer;
+import co.elastic.apm.agent.impl.Tracer;
+import co.elastic.apm.agent.impl.error.ErrorCapture;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+
+import javax.annotation.Nullable;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+import static co.elastic.apm.agent.loginstr.correlation.AbstractLogCorrelationHelper.ERROR_ID_MDC_KEY;
+import static co.elastic.apm.agent.loginstr.correlation.AbstractLogCorrelationHelper.TRACE_ID_MDC_KEY;
+import static co.elastic.apm.agent.loginstr.correlation.AbstractLogCorrelationHelper.TRANSACTION_ID_MDC_KEY;
+
+class CorrelationIdMapAdapter extends AbstractMap<String, String> {
+
+    private static final CorrelationIdMapAdapter INSTANCE = new CorrelationIdMapAdapter();
+    private static final Set<Entry<String, String>> ENTRY_SET = new TraceIdentifierEntrySet();
+    private static final List<String> ALL_KEYS = Arrays.asList(TRACE_ID_MDC_KEY, TRANSACTION_ID_MDC_KEY, ERROR_ID_MDC_KEY);
+    private static final Tracer tracer = GlobalTracer.get();
+    private static final List<Entry<String, String>> ENTRIES = Arrays.<Entry<String, String>>asList(
+        new LazyEntry(TRACE_ID_MDC_KEY, new Callable<String>() {
+            @Override
+            @Nullable
+            public String call() {
+                AbstractSpan<?> activeSpan = tracer.getActive();
+                if (activeSpan == null) {
+                    return null;
+                }
+                return activeSpan.getTraceContext().getTraceId().toString();
+            }
+        }),
+        new LazyEntry(TRANSACTION_ID_MDC_KEY, new Callable<String>() {
+            @Override
+            @Nullable
+            public String call() {
+                AbstractSpan<?> activeSpan = tracer.getActive();
+                if (activeSpan == null) {
+                    return null;
+                }
+                return activeSpan.getTraceContext().getId().toString();
```

Comment:
> This now sets the transaction.id to the active span's ID. I don't think that's what we want.

True!! I need to see again why the test did not cover that!
The intention was to do `activeSpan.getTraceContext().getTransactionId().toString()`, is getting the transaction object from the span preferable for any reason?

> Alternatively, we could change the implementation of currentTransaction() to return getActive().getTransaction() 

Makes sense, but I prefer doing it separately just in case there are implications we don't think about now

Reply:
> I need to see again why the test did not cover that!

Well, that's because the active span was the active transaction 🙂 
I'll add a test that creates and activates a span as well

Reply:
> The intention was to do activeSpan.getTraceContext().getTransactionId().toString(), is getting the transaction object from the span preferable for any reason?

I forgot that we had the transaction id directly within `TraceContext`.

> Makes sense, but I prefer doing it separately just in case there are implications we don't think about now

👌